doc/specs/stdlib_hash_maps.md

```diff
@@ -0,0 +1,3088 @@
+---
+title: Hash maps
+---
+
+# The `stdlib_32_bit_key_data_wrapper`, `stdlib_chaining_hash_map` and `stdlib_open_hash_map` modules
+
+(TOC)
+
+## Overview of hash maps
+
+The comparison of lexical entities or other objects for equality
+can be computationally expensive.
+This cost is often reduced by computing a near unique integer value,
+termed a hash code, from the structure of the object using a procedure
+termed a hash function.
+Equality of hash codes is a necessary, but not sufficient, condition
+for the original objects to be equal. 
+As integer comparisons are very efficient, performing an initial
+comparison of hash codes and then performing a detailed comparison
+only if the hash codes are equal can improve performance.
+The hash codes, in turn, can be mapped to a smaller set of integers,
+that can be used as an index, termed a hash index, to a rank one
+array, often termed a hash table or hash map.
+This document discusses the hash maps in the library.
+
+## Licensing
+
+The Fortran Standard Library is distributed under the MIT License.
+However components of the library should be evaluated as to whether
+they are compatible with the MTI License.
+The current hash maps were inspired by an
+[implementation](http://chasewoerner.org/src/hasht/) of David
+Chase. While the code has been greatly modified from his
+implementation, he has give permission for the unrestricted use of
+his code.
+
+
+## The hash map modules
+
+The Fortran Standard Library provides three modules for the
+implementation of simple hash maps. These maps only accept hash
+functions with a single argument, the key, and that yield a 32 bit
+hash code, The modules will need to be modified to use hash functions
+with a different API. There are three modules:
+`stdlib_32_bit_key_data_wrapper`, `stdlib_chaining_hash_map` and
+`stdlib_open_hash_map`, corresponding to the files:
+`stdlib_32_bit_key_data_wrapper.f90`, `stdlib_chaining_hash_map.f90`,
+and `stdlib_open_hash_map.f90`. The module
+`stdlib_32_bit_key_data_wrapper` provides an interface to the 32 bit
+hash functions of the Standard Library module,
+`stdlib_32_bit_hash_functions`, and provides wrappers to some of the
+hash functions so that they no longer need to be supplied seeds. The
+module `stdlib_chaining_hash_map` defines a datatype,
+`chaining_hash_map_type`, implementing a simple separate chaining hash
+map noted more for its diagnostics than its performance. Finally the
+module, `stdlib_open_hash_map` defines a datatype,
+`open_hash_map_type`, implementing a simple open addressing hash
+map noted more for its diagnostics than its performance.
```

Comment:
So both of the hash maps are `noted more for its diagnostics than its performance` ? That's OK, just want to double check that is the intention.

Reply:
I think the performance is not baad. Generating the information used in the diagnostics is, I suspect, a small amount of overhead, but it is an overhead other tables will lack.